File: javascript/24-lexicographic-permutations.js
function swap(permutation, i, j) {
  const temp = permutation[i];
  permutation[i] = permutation[j];
  permutation[j] = temp;
}

function reverse(permutation, i, num) {
  let start = i;
  let end = permutation.length - 1;

  let swapIndex = null;
  let swapValue = Infinity;

  while (start <= end) {
    swap(permutation, start, end);

    if (permutation[start] > num && permutation[start] <= swapValue) {
      swapIndex = start;
      swapValue = permutation[start];
    }
    if (permutation[end] > num && permutation[end] <= swapValue) {
      swapIndex = end;
      swapValue = permutation[end];
    }

    start++;
    end--;
  }

  return swapIndex;
}

function nextPermutation(permutation) {
  let index = permutation.length - 1;
  while (permutation[index - 1] > permutation[index]) {
    index--;
  }

  const swapIndex = reverse(permutation, index, permutation[index - 1]);
  if (swapIndex) {
    swap(permutation, index - 1, swapIndex);
  }
}

function lexicographicPermutations(n) {
  const permutation = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
  for(let i = 0; i < n; i++) {
    nextPermutation(permutation);
  }

  return permutation.join("");
}

console.log(lexicographicPermutations(999999));

File: go/24-lexicographic-permutations.go
package main

import "fmt"
import "strings"

func swap(permutation []int, i int, j int) {
	temp := permutation[i]
	permutation[i] = permutation[j]
	permutation[j] = temp
}

func reverse(permutation []int, i int, num int) int {
	start := i
	end := len(permutation) - 1
	swapIndex := -1
	swapValue := 10

	for start <= end {
		swap(permutation, start, end)

		if permutation[start] > num && permutation[start] < swapValue {
			swapIndex = start
			swapValue = permutation[start]
		}

		if permutation[end] > num && permutation[end] < swapValue {
			swapIndex = end
			swapValue = permutation[end]
		}

		start++
		end--
	}

	return swapIndex
}

func nextPermutation(permutation []int) {
	i := len(permutation) - 1
	for permutation[i-1] > permutation[i] {
		i--
	}

	swapIndex := reverse(permutation, i, permutation[i-1])
	if swapIndex != -1 {
		swap(permutation, i-1, swapIndex)
	}
}

func lexicographicPermutations(n int) string {
	permutation := []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
	for i := 0; i < n; i++ {
		nextPermutation(permutation)
	}

	return strings.Trim(strings.Join(strings.Fields(fmt.Sprint(permutation)), ""), "[]")
}

func main() {
	fmt.Println(lexicographicPermutations(999999))
}
